pkg/iacinventory/github_repo.go

```diff
@@ -0,0 +1,184 @@
+		strings.HasSuffix(info.Name(), ".yaml") ||
+		strings.HasSuffix(info.Name(), ".yml") ||
+		strings.HasSuffix(info.Name(), ".template") {
+		f, err := os.Open(filepath.Clean(path))
+		if err != nil {
+			return fmt.Errorf("error opening file during CloudFormation analysis: %w", err)
+		}
+		scanner := bufio.NewScanner(f)
+		for scanner.Scan() {
+			if bytes.Contains(scanner.Bytes(), []byte("AWSTemplateFormatVersion")) {
+				if g.CloudformationDirs == nil {
+					g.CloudformationDirs = make(map[string]bool)
+				}
+				g.CloudformationDirs[strings.TrimPrefix(filepath.Dir(filepath.Clean(path)), g.dir+"/")] = true
+			}
+		}
+		f.Close()
+	}
+	return nil
+}
+
+// GetMaster fetches and extracts the tarball of the master branch.
+func (g *GithubRepo) GetMaster(ctx context.Context, c *http.Client, username string, oauthToken string) error {
+	if username == "" || oauthToken == "" {
+		return fmt.Errorf("error fetching repository: credentials are not set")
+	}
+	req, _ := http.NewRequestWithContext(ctx, "GET", "https://api.github.com/repos/"+g.FullName+"/tarball/master", nil)
+	req.Header.Set("Authorization", "token "+oauthToken)
+	req.Header.Set("Accept", "application/vnd.github.v3.raw")
+	resp, err := c.Do(req)
+	if err != nil {
+		return fmt.Errorf("error sending HTTP request: %w", err)
+	}
+	defer resp.Body.Close()
+
+	// we stuff the extracted tarball into a temporary directory
+	dir, err := ioutil.TempDir("", "soluble-iac-scan")
+	if err != nil {
+		return fmt.Errorf("error creating temporary directory: %w", err)
+	}
+	g.dir = dir
+	gunzip, err := gzip.NewReader(resp.Body)
+	if err != nil {
+		// If there is no repository data, we'll encounter a gzip error
+		if errors.Is(err, gzip.ErrHeader) {
+			return fmt.Errorf("repository has no contents")
+		}
+		return fmt.Errorf("error creating gzip reader for tarball: %w", err)
+	}
+	defer gunzip.Close()
+	t := tar.NewReader(gunzip)
+
+	// write the tarball'd files to disk in the temporary directory
+	for {
+		header, err := t.Next()
+		if errors.Is(err, io.EOF) {
+			break
+		}
+		if err != nil {
+			log.Errorf("error extracting file from tarball: %w", err)
+			break
+		}
+
+		// cut out the top level directory to place files directly in g.dir
+		outfile := filepath.Join(g.dir,
+			filepath.Clean(strings.Join(strings.Split(header.Name, string(filepath.Separator))[1:], string(filepath.Separator))))
+		switch header.Typeflag {
+		case tar.TypeDir:
+			if err := os.Mkdir(outfile, 0o755); err != nil {
+				if os.IsExist(err) {
+					continue
+				}
+				return fmt.Errorf("error creating directory for tarball extraction: %w", err)
+			}
+		case tar.TypeReg:
+			// cut out the outer directory from the filename to ensure we add the files to g.dir (and not g.dir/some-random-subdir)
+			f, err := os.Create(outfile)
+			if err != nil {
+				return fmt.Errorf("error creating file during tarball extraction: %w", err)
+			}
+			c, err := io.CopyN(f, t, 100<<(10*2)) // 100MB max size
```

Comment:
This seems, big?  Surely we'd not be interested in files larger than even 1MB?

If refactoring this as a function this value should be a parameter or option to that function.

Reply:
My statement above was incorrect. Changed to limit single file size to 1MB.